### DISCUSSION TITLE
[Share] Add a Hyperlink to an annotation and jump to the target annotation

### DISCUSSION THREAD

**Discussion by YaoLiMuMu** [windingwind/zotero-actions-tags#296] [Action Scripts]
### Description

[Chinese] 这是我近两周尝试后得出的结果, 该脚本功能能够实现了两个PDF文件之间的字符级别的精准链接与定位. 欢迎各位大佬的指点与改善(在此由衷感谢@windingwind, @volatile-static, @northword三位前辈的帮助)
环境要求: Zotero 7,  请记住需要在actions&tags中添加两份独立的配置, 脚本代码参考Data
首先打开两份需要链接的文件, 在这里把它们分别称呼为目标PDF和原始PDF, 实现链接步骤如下:
1. 在目标PDF中选择要定位的地方添加高亮批注, 然后选中这个批注, 按下Ctrl+c按键复制这个批注到剪切板.
2. 在原始PDF中要添加链接的地方添加高亮批注, 然后选中这个批注, 鼠标右击在下拉的菜单栏中选择"Add Link"(这个动作会从系统剪切板中解析到目标PDF的批注信息).
3. 完成以上2个步骤, 就表示链接完成, 然后在原始PDF中的选中批注, 鼠标右击在下拉的菜单栏中选择"Goto Link", 就会跳转到目标PDF中了.

因为zotero好像与peek冲突了, 导致无法截取动态gif图, 用ffmpeg生成的gif素材文件会超过10M, 无法上传到github, 所以具体的效果图麻烦后续参与人员帮忙上传展示.

[English] Here is the result of my attempts in the past two weeks. This feature enables precise linking and positioning at the character level between two PDF files. Welcome to all experts for guidance and improvement (sincere thanks to @windingwind, @volatile-static, @northword).
Environment requirements: Zotero 7. Please remember to add two separate configurations in actions & tags, with script code reference in *Data*.
First, open the two PDF files that need to be linked. Here, they are referred to as the target PDF and the original PDF, respectively. The steps to create links are as follows:
1. In the target PDF, select the location where you want to add a highlight annotation, then select this annotation, press Ctrl+C to copy it to the clipboard.
2. In the original PDF, add an highlight annotation at the location where you want to add the link, then select this annotation, right-click and choose "Add Link" from the dropdown menu (this action will parse the annotation information from the system clipboard to the target PDF).
3. After completing the above two steps, the link is established. Then, select the annotation in the original PDF, right-click and choose "Goto Link" from the dropdown menu to jump to the target PDF.

Due to a conflict between Zotero and Peek, it is not possible to capture dynamic GIF images in my system. GIF material generated by ffmpeg exceeds 10MB and cannot be uploaded to GitHub. Therefore, please ask the participants to help upload specific effect images later on.

### Event

None

### Operation

Script

### Data For "Add Link"

```javascript
// @author yaolimumu
// @link https://github.com/windingwind/zotero-actions-tags/discussions/296
// @usage Add a Hyperlink to an annotation and jump to the target annotation
// @update 
// @note This is Add Link code
// @note This is Add Link code
// @note This is Add Link code

// Only execute once for all selected items
if (item) return;
item = items[0];
// EDIT THESE SETTINGS

const Zotero = require("Zotero");
const navigator = require("navigator");

function extractZoteroInfo(inputString) {
    let startIndex = inputString.indexOf("zotero://open-pdf/library/items/");
    let endIndex3 = inputString.lastIndexOf("))");
    let zoterolink = inputString.substring(startIndex, endIndex3);
    return zoterolink;
}

let oldlink = item.getRelations();
try {
  if (oldlink["dc:relation"][0] !== undefined) {
	item.removeRelation("dc:relation", oldlink["dc:relation"][0]);
}
} catch (error) { };

let copytext = await navigator.clipboard.readText();
let referInfo = extractZoteroInfo(copytext);

item.addRelation("dc:relation", referInfo);
await item.saveTx();
```
### Data for "Goto Link"
```javascript

// @author yaolimumu
// @link https://github.com/windingwind/zotero-actions-tags/discussions/296
// @usage Add a Hyperlink to an annotation and jump to the target annotation
// @update 
// @note This is Goto Link code
// @note This is Goto Link code
// @note This is Goto Link code

if(item) return;
item = items[0];

const Zotero = require("Zotero");

function extractZoteroInfo(inputString) {
    let startIndex = inputString.indexOf("zotero://open-pdf/library/items/") + "zotero://open-pdf/library/items/".length;
    let endIndex1 = inputString.indexOf("?page=");
    let endIndex2 = inputString.lastIndexOf("&annotation=");
    let endIndex3 = inputString.length;
    let zoteroId = inputString.substring(startIndex, endIndex1);
    let page = inputString.substring(endIndex1 + "?page=".length, endIndex2);
    let annotation = inputString.substring(endIndex2 + "&annotation=".length, endIndex3);
    return [zoteroId, page, annotation];
}

let copytext = item.getRelations()["dc:relation"][0];
let referInfo = extractZoteroInfo(copytext);

const userLibraryID = Zotero.Libraries.userLibraryID;
let getitem = Zotero.Items.getByLibraryAndKey(userLibraryID, referInfo[0]);
Zotero.FileHandlers.open(getitem, {
    location: {
        annotationID: referInfo[2],
        pageIndex: referInfo[1]
    }});
```

### Anything else

Add Link config
![image](https://github.com/windingwind/zotero-actions-tags/assets/32826811/8fe98ed4-4fe5-40c8-b7a9-d5546edd2e69)
Goto Link config
![image](https://github.com/windingwind/zotero-actions-tags/assets/32826811/dfc2be91-d070-4c83-8a69-34dae01cfcab)
Annotation Menu
![image](https://github.com/windingwind/zotero-actions-tags/assets/32826811/a3545a2c-dc1e-4be7-8eda-3f7a99edbac2)

**Comment by Geo123abc** [comment]
您好，非常好的功能，但是我好奇，能不能稍微介绍一下原理啊？我看添加链接之后，卡片中并不能见到链接，也没有一个标识，或者说创建的链接是“不可视”的吗？

**Reply by YaoLiMuMu** [reply]
对, 直接用脚本实现的, 新手练习, 链接目前是隐藏的(用户层不可见的), 因为作用是在阅读时快捷跳转两个相关联的文档使用, 我一般把这类链接用专门的批注高亮颜色, 与其他注释型批注作区分, 
基本原理是把目标批注的的链接数据(pdf的itemid, 批注key, 页参数等)复制到系统剪切板, 然后在源批注解析系统剪切板的链接信息, 保存在批注的元数据Relations字段上, 后面源批注的链接跳转就是解析自己的Relation字段上面的信息, 然后调用系统的打开文件函数实现跳转的.
另外. 批注在实现对象类型上也是属于条目item, 如果你想要要在用户层可见, 那么只能在批注的注释上面添加信息了(这样也会导致批注生成的笔记也会有链接信息), 元数据的字段是`annotationComment`, 以下是批注元信息的json格式例子, 希望能帮助到你.
```json
[
    "0": {
    "key": "CZZ42XQU",
    "version": 2706,
    "itemType": "annotation",
    "parentItem": "255HN5RN",
    "annotationType": "highlight",
    "annotationAuthorName": "",
    "annotationText": "参考５．１．４．３ 的电位均衡",
    "annotationComment": "",
    "annotationColor": "#ffd400",
    "annotationPageLabel": "9",
    "annotationSortIndex": "00008|000230|00446",
    "annotationPosition": "{\"pageIndex\":8,\"rects\":[[68.604,376.224,174.372,395.527]]}",
    "tags": [],
    "relations": {},
    "dateAdded": "2024-03-12T10:35:00Z",
    "dateModified": "2024-03-12T10:35:00Z"
}
]
```

**Reply by Geo123abc** [reply]
感谢回复，这样看来，zotero的annotation有很多可以开发的功能啊

**Reply by Geo123abc** [reply]
https://github.com/zzlb0224/zotero-annotation-manage/releases/tag/v0.4.28

**Comment by zzlb0224** [comment]
非常好的想法，在zotero里面实现了链接功能。
学习一下

**Reply by YaoLiMuMu** [reply]
十分乐意能方便你的zotero使用, 改善你的文献阅读体验.

**Comment by Zane-Zzx** [comment]
请问link后似乎同步出了问题，显示 Unsupported predicate 'link:annotation' 是什么原因？
除此之外，直接删掉批注是不是就是清除干净了，相当于Delete link了

**Reply by YaoLiMuMu** [reply]
已调整修改.
因为"link:annotation"是自定义的key, zotero云同步时识别不了该key, 可以改键 "link:annotation" 改成 "dc:relation", 这样同步就不会有问题了
问题二: Yes, Link信息是添加到批注relations字段的, 直接删掉批注, link也就没有了.
PS: 嫌麻烦的话, 直接用这个插件 [zotero-annotation-manage](https://github.com/zzlb0224/zotero-annotation-manage/releases/tag/v0.4.28)也实现了批注链接功能, 同时也添加了一些UI(如预览关联链接和删除操作)

**Comment by Bentmj** [comment]
'relations' values currently must be Zotero item URIs
请问你还有在用这个脚本吗 目前同步会遇到这个反馈 该如何修改呢
